BUG: Use after free in `CommandFetchFile` reported by ASan

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.


### Version

unstable

### Minimal reproduce step


```
./build.sh build -DENABLE_ASAN=ON
cd tests/tcl
./runtest --dont-clean
grep Sanitizer tests/tmp -r
```

### What did you expect to see?

No ASAN reports

### What did you see instead?

For more detail can see: https://github.com/apache/incubator-kvrocks/runs/7174515538?check_suite_focus=true

```
\o/ All tests passed without errors!
tests/tmp/server.21550.23/stderr:==30422==ERROR: AddressSanitizer: heap-use-after-free on address 0x606000003500 at pc 0x7f97aad45733 bp 0x7f97961e5840 sp 0x7f97961e4fe8
tests/tmp/server.21550.23/stderr:SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79732) 
\ndetail reports:\n
E20220704 05:25:13.181370 29782 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:14.181675 29782 replication.cc:248] [replication] Failed to connect the master, err: Connection refused
E20220704 05:25:59.096598  1453 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:26:01.526546  1765 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:40.944598 31108 slot_migrate.cc:476] [migrate] Failed to read response, Err: Connection reset by peer
E20220704 05:25:40.944772 31108 slot_migrate.cc:792] [migrate] Wrong response
E20220704 05:25:40.944813 31108 slot_migrate.cc:588] [migrate] Failed to migrate complex key: 0QG
E20220704 05:25:40.944831 31108 slot_migrate.cc:310] [migrate] Failed to migrate key: 0QG
E20220704 05:25:40.944859 31108 slot_migrate.cc:181] [migrate] Failed to send snapshot
E20220704 05:25:40.944891 31108 slot_migrate.cc:420] [migrate] Failed to send import command to destination, slot: 8, error: Broken pipe
E20220704 05:24:59.634393 26964 replication.cc:77] Ping slave[127.0.0.1:41322] err: Broken pipe, would stop the thread
WARNING: Logging before InitGoogleLogging() is written to STDERR
I20220704 05:25:50.447059 21587 event_listener.cc:132] [event_listener/table_file_deleted] db: ./tests/tmp/server.21553.1/db, sst file: ./tests/tmp/server.21553.1/db/000017.sst, status: OK
=================================================================
==30422==ERROR: AddressSanitizer: heap-use-after-free on address 0x606000003500 at pc 0x7f97aad45733 bp 0x7f97961e5840 sp 0x7f97961e4fe8
READ of size 38 at 0x606000003500 thread T103 (feed-repl-file)
    #0 0x7f97aad45732  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79732)
    #1 0x5560df5a4ece in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaabece)
    #2 0x5560df5a4f0e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaabf0e)
    #3 0x5560df048ff1 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > std::operator+<char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*) /usr/include/c++/7/bits/basic_string.h:5928
    #4 0x5560df012557 in Engine::Storage::ReplDataManager::OpenDataFile(Engine::Storage*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long*) /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/storage.cc:836
    #5 0x5560dedaf46d in Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}::operator()() const /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/redis_cmd.cc:4395
    #6 0x5560dedb0a3c in void std::__invoke_impl<void, Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}>(std::__invoke_other, Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}&&) /usr/include/c++/7/bits/invoke.h:60
    #7 0x5560dedb0a3c in std::__invoke_result<Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}>::type std::__invoke<Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}>(std::__invoke_result&&, (Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}&&)...) /usr/include/c++/7/bits/invoke.h:95
    #8 0x5560dedb0a3c in decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/include/c++/7/thread:234
    #9 0x5560dedb0a3c in std::thread::_Invoker<std::tuple<Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}> >::operator()() /usr/include/c++/7/thread:243
    #10 0x5560dedb0a3c in std::thread::_State_impl<std::thread::_Invoker<std::tuple<Redis::CommandFetchFile::Execute(Server*, Redis::Connection*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)::{lambda()#1}> > >::_M_run() /usr/include/c++/7/thread:186
    #11 0x5560df5a07de in execute_native_thread_routine (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaa77de)
    #12 0x7f97aaab46da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)
    #13 0x7f97aa43f61e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x12161e)
0x606000003500 is located 0 bytes inside of 55-byte region [0x606000003500,0x606000003537)
freed by thread T0 here:
    #0 0x7f97aadad2c0 in operator delete(void*) (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xe12c0)
    #1 0x5560debe9a12 in __gnu_cxx::new_allocator<char>::deallocate(char*, unsigned long) /usr/include/c++/7/ext/new_allocator.h:125
    #2 0x5560debe9a12 in std::allocator_traits<std::allocator<char> >::deallocate(std::allocator<char>&, char*, unsigned long) /usr/include/c++/7/bits/alloc_traits.h:462
    #3 0x5560debe9a12 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy(unsigned long) /usr/include/c++/7/bits/basic_string.h:226
    #4 0x5560debe9a12 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose() /usr/include/c++/7/bits/basic_string.h:221
    #5 0x5560debe9a12 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() /usr/include/c++/7/bits/basic_string.h:647
    #6 0x5560debe9a12 in Config::~Config() /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/config.h:66
    #7 0x5560deb9b44c in main /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/main.cc:316
    #8 0x7f97aa33fc86 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21c86)
previously allocated by thread T0 here:
    #0 0x7f97aadac448 in operator new(unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xe0448)
    #1 0x5560df5a28f9 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_mutate(unsigned long, unsigned long, char const*, unsigned long) (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaa98f9)
Thread T103 (feed-repl-file) created by T97 (worker) here:
    #0 0x7f97aad03d2f in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x37d2f)
    #1 0x5560df5a0ac4 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaa7ac4)
    #2 0x5560dee1f8a4 in Redis::Connection::ExecuteCommands(std::vector<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&) /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/redis_connection.cc:422
    #3 0x5560dee233f5 in Redis::Connection::OnRead(bufferevent*, void*) /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/redis_connection.cc:89
    #4 0x5560df4df61d in bufferevent_run_deferred_callbacks_unlocked /home/runner/work/incubator-kvrocks/incubator-kvrocks/build/_deps/libevent-src/bufferevent.c:208
Thread T97 (worker) created by T0 here:
    #0 0x7f97aad03d2f in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x37d2f)
    #1 0x5560df5a0ac4 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/tcl/redis-server+0xaa7ac4)
    #2 0x5560def939bb in Server::Start() /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/server.cc:121
    #3 0x5560deb9b2ee in main /home/runner/work/incubator-kvrocks/incubator-kvrocks/src/main.cc:354
    #4 0x7f97aa33fc86 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21c86)
SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79732) 
Shadow bytes around the buggy address:
  0x0c0c7fff8650: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8660: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8670: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8680: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fa
  0x0c0c7fff8690: fa fa fa fa fd fd fd fd fd fd fd fa fa fa fa fa
=>0x0c0c7fff86a0:[fd]fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
  0x0c0c7fff86b0: fd fd fd fa fa fa fa fa fd fd fd fd fd fd fd fa
  0x0c0c7fff86c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff86d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff86e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff86f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==30422==ABORTING
E20220704 05:25:13.181382 29785 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:24:53.779562 25091 replication.cc:128] Write error while sending batch to slave: Broken pipe. batches: 0x2433380D0A02000000000000000100000003013105010D0B5F5F6E616D65737061636561060100000000620D0A
E20220704 05:24:53.780030 25099 replication.cc:128] Write error while sending batch to slave: Broken pipe. batches: 0x2433380D0A02000000000000000100000003013105010D0B5F5F6E616D65737061636561060100000000620D0A
E20220704 05:24:53.780268 25131 replication.cc:128] Write error while sending batch to slave: Broken pipe. batches: 0x2433380D0A02000000000000000100000003013105010D0B5F5F6E616D65737061636561060100000000620D0A
E20220704 05:24:56.222230 25861 slot_migrate.cc:241] [migrate] Failed to connect destination server
E20220704 05:24:56.222508 25861 slot_migrate.cc:170] [migrate] Failed to start migrating
E20220704 05:24:49.706113 24751 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:52.255283 32602 slot_migrate.cc:288] [migrate] Stop migrating snapshot due to the thread stoped
E20220704 05:25:52.255439 32602 slot_migrate.cc:181] [migrate] Failed to send snapshot
E20220704 05:25:04.225044 26691 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:56.253038   591 slot_migrate.cc:494] [migrate] Got invalid response: -NOAUTH Authentication required., line length: 32
E20220704 05:25:56.253249   591 slot_migrate.cc:256] [migrate] Failed to notify the destination to prepare to import data
E20220704 05:25:56.253271   591 slot_migrate.cc:170] [migrate] Failed to start migrating
E20220704 05:25:56.253386   591 slot_migrate.cc:494] [migrate] Got invalid response: -NOAUTH Authentication required., line length: 32
E20220704 05:25:56.254267   591 slot_migrate.cc:494] [migrate] Got invalid response: -ERR invalid password, line length: 21
E20220704 05:25:56.254307   591 slot_migrate.cc:406] [migrate] Failed to auth destination server with 'pass', stop migrating slot 22
E20220704 05:25:56.254325   591 slot_migrate.cc:170] [migrate] Failed to start migrating
E20220704 05:25:56.254431   591 slot_migrate.cc:494] [migrate] Got invalid response: -NOAUTH Authentication required., line length: 32
E20220704 05:25:10.059321 29194 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:06.633349 28188 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:24:59.645223 27469 replication.cc:248] [replication] Failed to connect the master, err: Connection refused
E20220704 05:25:02.869042 27470 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:24:46.901080 23284 redis_cmd.cc:1582] Failed to execute redis command: blpop, err: Invalid argument: WRONGTYPE Operation against a key holding the wrong kind of value
E20220704 05:24:48.914759 23284 redis_cmd.cc:1582] Failed to execute redis command: brpop, err: Invalid argument: WRONGTYPE Operation against a key holding the wrong kind of value
E20220704 05:25:41.277627 30700 replication.cc:698] [replication] Failed to parallel fetch files while fetch file err: read size: File exists
E20220704 05:25:51.290735 30700 replication.cc:162] [replication] connection error/eof, reconnect the master
E20220704 05:25:48.578738 31962 slot_migrate.cc:494] [migrate] Got invalid response: -MOVED 10 127.0.0.1:28122, line length: 25
E20220704 05:25:48.5[790](https://github.com/apache/incubator-kvrocks/runs/7174515538?check_suite_focus=true#step:9:791)25 31962 slot_migrate.cc:792] [migrate] Wrong response
E20220704 05:25:48.579075 31962 slot_migrate.cc:588] [migrate] Failed to migrate complex key: 7Ok
E20220704 05:25:48.5[791](https://github.com/apache/incubator-kvrocks/runs/7174515538?check_suite_focus=true#step:9:792)13 31962 slot_migrate.cc:310] [migrate] Failed to migrate key: 7Ok
E20220704 05:25:48.57[913](https://github.com/apache/incubator-kvrocks/runs/7174515538?check_suite_focus=true#step:9:914)8 31962 slot_migrate.cc:181] [migrate] Failed to send snapshot
sanitizer error was reported, exiting...
Error: Process completed with exit code 1.
```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/src/main.cc
+++ b/src/main.cc
@@ -358,6 +358,7 @@ int main(int argc, char* argv[]) {
   }
   srv->Join();
 
+  delete srv;
   removePidFile(config.pidfile);
   google::ShutdownGoogleLogging();
   libevent_global_shutdown();

--- a/src/redis_cmd.cc
+++ b/src/redis_cmd.cc
@@ -1577,7 +1577,7 @@ class CommandBPop : public Commander {
       timeval tm = {timeout_, 0};
       evtimer_add(timer_, &tm);
     }
-    return Status::OK();
+    return Status(Status::BlockingCmd);
   }
 
   rocksdb::Status TryPopFromList() {
@@ -3362,7 +3362,7 @@ class CommandExec : public Commander {
     }
 
     // Reply multi length first
-    conn->Reply(Redis::MultiLen(conn->GetMultiExecCommands().size()));
+    conn->Reply(Redis::MultiLen(conn->GetMultiExecCommands()->size()));
     // Execute multi-exec commands
     conn->SetInExec();
     conn->ExecuteCommands(conn->GetMultiExecCommands());

--- a/src/redis_cmd.cc
+++ b/src/redis_cmd.cc
@@ -4401,6 +4401,7 @@ class CommandFetchFile : public Commander {
       svr->IncrFetchFileThread();
 
       for (auto file : files) {
+        if (svr->IsStopped()) break;
         uint64_t file_size = 0, max_replication_bytes = 0;
         if (svr->GetConfig()->max_replication_mb > 0) {
           max_replication_bytes = (svr->GetConfig()->max_replication_mb*MiB) /

--- a/src/redis_connection.cc
+++ b/src/redis_connection.cc
@@ -87,7 +87,6 @@ void Connection::OnRead(struct bufferevent *bev, void *ctx) {
     return;
   }
   conn->ExecuteCommands(conn->req_.GetCommands());
-  conn->req_.ClearCommands();
   if (conn->IsFlagEnabled(kCloseAsync)) {
     conn->Close();
   }
@@ -307,13 +306,14 @@ void Connection::recordProfilingSampleIfNeed(const std::string &cmd, uint64_t du
   svr_->GetPerfLog()->PushEntry(entry);
 }
 
-void Connection::ExecuteCommands(const std::vector<Redis::CommandTokens> &to_process_cmds) {
-  if (to_process_cmds.empty()) return;
-
+void Connection::ExecuteCommands(std::list<Redis::CommandTokens> *to_process_cmds) {
   Config *config = svr_->GetConfig();
   std::string reply, password = config->requirepass;
 
-  for (auto &cmd_tokens : to_process_cmds) {
+  while (!to_process_cmds->empty()) {
+    auto cmd_tokens = to_process_cmds->front();
+    to_process_cmds->pop_front();
+
     if (IsFlagEnabled(Redis::Connection::kCloseAfterReply) &&
         !IsFlagEnabled(Connection::kMultiExec)) break;
     if (GetNamespace().empty()) {
@@ -426,6 +426,12 @@ void Connection::ExecuteCommands(const std::vector<Redis::CommandTokens> &to_pro
     svr_->SlowlogPushEntryIfNeeded(current_cmd_->Args(), duration);
     svr_->stats_.IncrLatency(static_cast<uint64_t>(duration), cmd_name);
     svr_->FeedMonitorConns(this, cmd_tokens);
+
+    // Break the execution loop when occurring the blocking command like BLPOP or BRPOP,
+    // it will suspend the connection and wait for the wakeup signal.
+    if (s.IsBlockingCommand()) {
+      break;
+    }
     // Reply for MULTI
     if (!s.IsOK()) {
       Reply(Redis::Error("ERR " + s.Msg()));

--- a/src/redis_connection.cc
+++ b/src/redis_connection.cc
@@ -306,7 +306,7 @@ void Connection::recordProfilingSampleIfNeed(const std::string &cmd, uint64_t du
   svr_->GetPerfLog()->PushEntry(entry);
 }
 
-void Connection::ExecuteCommands(std::list<Redis::CommandTokens> *to_process_cmds) {
+void Connection::ExecuteCommands(std::deque<CommandTokens> *to_process_cmds) {
   Config *config = svr_->GetConfig();
   std::string reply, password = config->requirepass;
 

--- a/src/redis_connection.h
+++ b/src/redis_connection.h
@@ -21,6 +21,7 @@
 #pragma once
 
 #include <vector>
+#include <list>
 #include <string>
 #include <utility>
 #include <memory>
@@ -102,7 +103,7 @@ class Connection {
   evbuffer *Input() { return bufferevent_get_input(bev_); }
   evbuffer *Output() { return bufferevent_get_output(bev_); }
   bufferevent *GetBufferEvent() { return bev_; }
-  void ExecuteCommands(const std::vector<Redis::CommandTokens> &to_process_cmds);
+  void ExecuteCommands(std::list<Redis::CommandTokens> *to_process_cmds);
   bool isProfilingEnabled(const std::string &cmd);
   void recordProfilingSampleIfNeed(const std::string &cmd, uint64_t duration);
   void SetImporting() { importing_ = true; }
@@ -113,7 +114,7 @@ class Connection {
   bool IsInExec() { return in_exec_; }
   bool IsMultiError() { return multi_error_; }
   void ResetMultiExec();
-  const std::vector<Redis::CommandTokens> &GetMultiExecCommands() { return multi_cmds_; }
+  std::list<Redis::CommandTokens> *GetMultiExecCommands() { return &multi_cmds_; }
 
   std::unique_ptr<Commander> current_cmd_;
   std::function<void(int)> close_cb_ = nullptr;
@@ -142,7 +143,7 @@ class Connection {
   Server *svr_;
   bool in_exec_ = false;
   bool multi_error_ = false;
-  std::vector<Redis::CommandTokens> multi_cmds_;
+  std::list<Redis::CommandTokens> multi_cmds_;
 
   bool importing_ = false;
 };

--- a/src/redis_connection.h
+++ b/src/redis_connection.h
@@ -21,7 +21,7 @@
 #pragma once
 
 #include <vector>
-#include <list>
+#include <deque>
 #include <string>
 #include <utility>
 #include <memory>
@@ -103,7 +103,7 @@ class Connection {
   evbuffer *Input() { return bufferevent_get_input(bev_); }
   evbuffer *Output() { return bufferevent_get_output(bev_); }
   bufferevent *GetBufferEvent() { return bev_; }
-  void ExecuteCommands(std::list<Redis::CommandTokens> *to_process_cmds);
+  void ExecuteCommands(std::deque<CommandTokens> *to_process_cmds);
   bool isProfilingEnabled(const std::string &cmd);
   void recordProfilingSampleIfNeed(const std::string &cmd, uint64_t duration);
   void SetImporting() { importing_ = true; }
@@ -114,7 +114,7 @@ class Connection {
   bool IsInExec() { return in_exec_; }
   bool IsMultiError() { return multi_error_; }
   void ResetMultiExec();
-  std::list<Redis::CommandTokens> *GetMultiExecCommands() { return &multi_cmds_; }
+  std::deque<Redis::CommandTokens> *GetMultiExecCommands() { return &multi_cmds_; }
 
   std::unique_ptr<Commander> current_cmd_;
   std::function<void(int)> close_cb_ = nullptr;
@@ -143,7 +143,7 @@ class Connection {
   Server *svr_;
   bool in_exec_ = false;
   bool multi_error_ = false;
-  std::list<Redis::CommandTokens> multi_cmds_;
+  std::deque<Redis::CommandTokens> multi_cmds_;
 
   bool importing_ = false;
 };

--- a/src/redis_request.h
+++ b/src/redis_request.h
@@ -20,6 +20,7 @@
 
 #pragma once
 
+#include <list>
 #include <vector>
 #include <string>
 #include <event2/buffer.h>
@@ -44,8 +45,7 @@ class Request {
   // Parse the redis requests (bulk string array format)
   Status Tokenize(evbuffer *input);
 
-  const std::vector<CommandTokens> &GetCommands() { return commands_; }
-  void ClearCommands() { commands_.clear(); }
+  std::list<CommandTokens> *GetCommands() { return &commands_; }
 
  private:
   // internal states related to parsing
@@ -55,7 +55,7 @@ class Request {
   int64_t multi_bulk_len_ = 0;
   size_t bulk_len_ = 0;
   CommandTokens tokens_;
-  std::vector<CommandTokens> commands_;
+  std::list<CommandTokens> commands_;
 
   Server *svr_;
 };

--- a/src/redis_request.h
+++ b/src/redis_request.h
@@ -20,7 +20,7 @@
 
 #pragma once
 
-#include <list>
+#include <deque>
 #include <vector>
 #include <string>
 #include <event2/buffer.h>
@@ -45,7 +45,7 @@ class Request {
   // Parse the redis requests (bulk string array format)
   Status Tokenize(evbuffer *input);
 
-  std::list<CommandTokens> *GetCommands() { return &commands_; }
+  std::deque<CommandTokens> *GetCommands() { return &commands_; }
 
  private:
   // internal states related to parsing
@@ -55,7 +55,7 @@ class Request {
   int64_t multi_bulk_len_ = 0;
   size_t bulk_len_ = 0;
   CommandTokens tokens_;
-  std::list<CommandTokens> commands_;
+  std::deque<CommandTokens> commands_;
 
   Server *svr_;
 };

--- a/src/scripting.cc
+++ b/src/scripting.cc
@@ -90,6 +90,7 @@ namespace Lua {
   }
 
   void DestroyState(lua_State *lua) {
+    lua_gc(lua, LUA_GCCOLLECT, 0);
     lua_close(lua);
   }
 

--- a/src/server.cc
+++ b/src/server.cc
@@ -81,6 +81,19 @@ Server::~Server() {
   for (const auto &iter : conn_ctxs_) {
     delete iter.first;
   }
+
+  // Wait for all fetch file threads stop and exit and force destroy
+  // the server after 60s.
+  int counter = 0;
+  while (GetFetchFileThreadNum() != 0) {
+    usleep(100000);
+    if (++counter == 600) {
+      LOG(WARNING) << "Will force destroy the server after waiting 60s, leave " << GetFetchFileThreadNum()
+                   << " fetch file threads are still running";
+      break;
+    }
+  }
+  Lua::DestroyState(lua_);
 }
 
 // Kvrocks threads list:

--- a/src/status.h
+++ b/src/status.h
@@ -55,6 +55,9 @@ class Status {
 
     // Network
     NetSendErr,
+
+    // Blocking
+    BlockingCmd,
   };
 
   Status() : Status(cOK) {}
@@ -62,6 +65,7 @@ class Status {
   bool IsOK() { return code_ == cOK; }
   bool IsNotFound() { return code_ == NotFound; }
   bool IsImorting() { return code_ == SlotImport; }
+  bool IsBlockingCommand() { return code_ == BlockingCmd; }
   std::string Msg() {
     if (IsOK()) {
       return "ok";

--- a/src/worker.cc
+++ b/src/worker.cc
@@ -62,6 +62,9 @@ Worker::~Worker() {
   for (const auto &iter : conns_) {
     conns.emplace_back(iter.second);
   }
+  for (const auto &iter : monitor_conns_) {
+    conns.emplace_back(iter.second);
+  }
   for (const auto &iter : conns) {
     iter->Close();
   }

